src/platform/android/java/chip/platform/NsdManagerServiceResolver.java

```diff
@@ -223,4 +255,47 @@ public void removeServices() {
     registrationListeners.clear();
     mMFServiceName.clear();
   }
+
+  /**
+   * The Android NsdManager calls back on the NsdManager.ResolveListener with a
+   * FAILURE_ALREADY_ACTIVE(3) if any application code calls resolveService() on it while the
+   * resolve operation is already active (from another call made previously). An object of
+   * NsdManagerResolverAvailState allows NsdManagerServiceResolver to synchronize on the usage of
+   * NsdManager's resolveService() API
+   */
+  public static class NsdManagerResolverAvailState {
+    private static final String TAG = NsdManagerResolverAvailState.class.getSimpleName();
+
+    private Lock lock = new ReentrantLock();
+    private Condition condition = lock.newCondition();
+    private boolean busy = false;
+
+    /**
+     * Waits if the NsdManager is already busy with resolving a service. Otherwise, it marks it as
+     * busy and returns
+     */
+    public void markBusyIfFreeOrWait() {
+      lock.lock();
+      try {
+        while (busy) {
+          Log.d(TAG, "Found NsdManager Resolver busy, waiting");
+          condition.await();
+        }
+        Log.d(TAG, "Found NsdManager Resolver free, using it");
+        busy = true;
+      } catch (InterruptedException e) {
+        Log.e(TAG, "Failure while waiting for condition: " + e);
+      }
+      lock.unlock();
+    }
+
+    /** Signals the NsdManager resolver as free */
+    public void signalFree() {
+      lock.lock();
+      Log.d(TAG, "Signaling NsdManager Resolver as free");
+      busy = false;
+      condition.signalAll();
```

Comment:
would `signal` be sufficient here? Why the All?

Reply:
signal is indeed more appropriate here. Changed.